# Welcome to Lab: Similarity 🧍🧍‍♂️🧍‍♀️

Many of you already know the people you work with in your lab. However, have you ever wondered if there is someone even more similar to you than the people you already know?

In this lab, we'll once again explore the dataset from the Hello Survey you've completed, utilizing some new Python skills to calculate who is most similar to you!

A few tips to remember:

- Refer to your **lecture notebook** and the **pandas cheat sheet** to help you out with the code!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask!

Let's get started!

In [ ]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Daniel"
ca1_name = "Linh"
ca2_name = "Eeshan"


# Say hello to each other!
# - Groups of 3 are ideal :)
# - However, groups of 2 or 4 are fine too!

# QOTD to ask your group: What's your favorite (or least favorite) building on campus? 
partner1_name = "David Mocianko"
partner1_netid = "dmoci2"
partner1_favbuilding = "CIF"

partner2_name = "Shane Moncayl"
partner2_netid = "moncayo3"
partner2_favbuilding = "CIF"

partner3_name = "Evan Walker"
partner3_netid = "evanaw3"
partner3_favbuilding = "CIF"

<hr style="color: #DD3403;">

# Part 1: Hello, Again

## Load the Hello Dataset
To find those most similar to you in this course, we're going to need the good old Hello Dataset.


The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-fa22.csv
```

Use Python to load this dataset into a DataFrame called `df_hello`:

In [6]:
import pandas as pd
df_hello = pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-fa22.csv")
df_hello

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0     CA Ana  AL1                    2.0                    5.0   
1     CA Ram  AYR                    2.0                    5.0   
2     CA Xin  AYO                    1.0                    5.0   
3      Lekha  AYH                    2.0                    2.0   
4       Reni  AYD                    2.0                    5.0   
..       ...  ...                    ...                    ...   
557    Emily  AYF                    2.0                    5.0   
558  Natalia  AYM                    1.0                    3.0   
559    Abrar  AY)                    1.0                    3.0   
560    Chris  AYA                    1.0                    3.0   
561     Jake  ATG                    1.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
4                                            2-Jan                 7UP   
..                                             ...                 ...   
557                                            0.5         Sierra Mist   
558                                            0.5         Sierra Mist   
559                                          2-Jan                 7UP   
560                                          2-Jan                 7UP   
561                                            0.5         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
4               iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
557             iOS            Apple          Beef   On/Before June 30  ...   
558             iOS            Apple          Beef   On/Before June 30  ...   
559             iOS           Orange          Beef     On/After July 1  ...   
560         Android           Orange          Beef     On/After July 1  ...   
561             iOS            Apple          Beef   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
4                                      Mocha   
..                                       ...   
557          Caramel Macchiato with oat milk   
558                                Green Tea   
559                               Chai latte   
560                      Vanilla Iced Coffee   
561                              just coffee   

    What is your favorite mobile app? What is your favorite subject?  \
0                            Snapchat                           Math   
1                      Google chrome                     Mathematics   
2                           Starbucks                     Psychology   
3                             twitter              english (writing)   
4                           Instagram                           Math   
..                                ...                            ...   
557                           Tik Tok                        History   
558                          Snapchat                           Math   
559                    Clash of Clans                        english   
560                         Instagram              Political Science

### 🔬 Test Case Checkpoint 🔬

In [5]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert("df_hello" in vars()), "This is not the Hello dataset you're looking for. Check the URL."
assert("Android or iOS?" in df_hello), "This is not the Hello dataset you're looking for. Check the URL."
assert(len(df_hello) >= 524), "This is not the Hello dataset you're looking for. Check the URL."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Finding Yourself
There is no finding people you are compatible with before finding yourself - literally!

### Puzzle 1.1: Your Row
In the following cell, write the Python code to create a DataFrame with only yourself:

In [7]:
df_me = df_hello[df_hello.Name=="Sabarish"]
df_me

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
119  Sabarish  AYD                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
119                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
119             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
119         Java Chip Frappacino - Starbucks   

    What is your favorite mobile app? What is your favorite subject?  \
119                         Instagram                           Math   

    Who is your favorite singer? What is your favorite season?  \
119                   Bruno Mars                          Fall   

    What is your Zodiac sign? What is your favorite video game?  \
119                     Virgo                             Forza   

    What is your favorite non-video game? What is your favorite food?  \
119                              Monopoly               Chicken Wings   

    What's your favorite movie?  
119         Wolf of Wall Street  

[1 rows x 120 columns]

## DataFrame vs. Series

Before we continue, a brief lesson on some of the datatypes you'll encounter in this class. 

In `pandas`, there are three data structures we will be using, two of which you have already seen:

- A **DataFrame** is similar to an Excel sheet. It contains zero or more rows and zero or more columns. It can contain up to an unlimited number of rows and columns and is the basic data structure we've worked with from the start

- A **GroupBy** is a grouping of rows within a DataFrame. This "three dimensional data" requires us to use `agg` to reduce it back down to a DataFrame - which you've been doing in homeworks and labs thus far!

- A **Series** is a single row with zero or more values (columns). You can think of a DataFrame as a collection of **Series** objects. When we deal with a single row, we will usually want to deal with a **Series** instead of a **one row DataFrame**

In this lab, we will learn how to convert from a **one row DataFrame** to a **Series**, but here's a quick recap on how to convert from types:

### Converting Types

The `pandas` library makes it easy to convert between a **Series**, **DataFrame**, and **GroupBy**.

- From a **Series** (`row`) to a **DataFrame**
    - `row.to_frame()`
- From a **DataFrame** (`df`) to a **GroupBy**:
    - `df.groupby(by='columnName')` -- (You know this already!)
- From a **GroupBy** (`group`) to a **DataFrame**:
    - `df.agg( function )` -- (You know this already!)
- From a **DataFrame** (`df`) to a **Series**:
    - `df.iloc[0]`, if the DataFrame has one row

See this guide, https://discovery.cs.illinois.edu/guides/DataFrame-Fundamentals/dataframe-loc-vs-iloc/, for more information about `iloc[]`. 

### Puzzle 1.2: Transform your row into a Series
Using the correct function from above, transform the row containing only you from a one-row DataFrame into a Series. Make sure the Series is named my_row (this will match up with code we provide you with later):

In [8]:
my_row = df_me.iloc[0]

### 🔬 Test Case Checkpoint 🔬

In [9]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(type(my_row) == pd.core.series.Series), "Make sure you've correctly converted your one-row DataFrame into a Series..."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Choosing Your Columns

We have to think about our approach to finding your STAT107 twin before we begin. 

You may have noticed our Hello Dataset has over **100 questions**. Instead of finding the most similar person to you using **every question**, we'll pick out a few questions to analyze. 

**But first, we need to revisit the questions.**

Remember our  old friend `df.columns.tolist()`. To recap, this command will list all of the columns (variables) in a DataFrame.

 Use `df.columns.tolist()` in the cell below. 

You wil likely see the message: Output exceeds the __size limit__. Open the full output data __in a text editor__
If you do, click the in a text editor link to open the whole list in a new tab. :)


In [10]:
df_hello.columns.tolist()

['Name',
 'Lab',
 'Number 1 or Number 2?',
 'Number 3 or Number 5?',
 'What is your preferred way of saying one half?',
 '7UP or Sierra Mist?',
 'Android or iOS?',
 'Apple or Orange?',
 'Beef or Pork?',
 'Your birthday is...',
 'Do you prefer to wear black or white?',
 'Boba or Coffee?',
 'Cake or Pie?',
 'Canes or Chick-fil-a?',
 'Do you pour milk or cereal first?',
 'Do you prefer chocolate or vanilla?',
 'Coffee or Tea?',
 'Coke or Pepsi?',
 'Cookies or Ice Cream?',
 'Do you prefer ice cream in a cup or in a cone?',
 'Do you prefer dad or mom?',
 'Marvel or DC?',
 'Are you living in a dorm or an apartment?',
 'Your last digit in your NetID...',
 'Were you born in an even or odd month of the year?',
 'Would you rather have the superpower of flight or invisibility?',
 'UIUC Football or Basketball Team?',
 'French Fries or Tater Tots?',
 'Heads or Tails?',
 'Hot Weather or Cold Weather?',
 'Hot Dogs or Hamburgers?',
 'In-State or Out-of-State/International',
 'Introvert or Extrovert?',

### Puzzle 1.3: Choosing Your Columns
From the above list of columns (questions), choose at least **12 questions** to work with moving forward. Since there are about 120 total questions, we'll work with at least 10%. 

Store them as a **list** of **strings** named `questions` in the Python cell below:

In [11]:
questions = ["Android or iOS?","Canes or Chick-fil-a?","Marvel or DC?","UIUC Football or Basketball Team?","In-State or Out-of-State/International","Introvert or Extrovert?","Morning Person or Night Person?","Do you like partying at the University of Urbana-Champaign or do you want to go back home?","Rap Music or Pop Music?","Starbucks or Dunkin?","Spring or Fall?","TV Shows or Movies?"]

### 🔬 Test Case Checkpoint 🔬

In [12]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(questions) >= 12), "Do you have at least 12 questions in your columns list?"
for q in questions:
   assert(q in df_hello), "Double check that you've copied each question correctly"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Question Choice

**Q: Explain why you chose the questions you did to evaluate similarity. Is there a question or questions you think that say more about someone than others? Write at least 3 complete sentences.**

*(✏️ I chose the questions that I felt that I mostly valued and that defined my characteristics the most. I feel that if I am trying to find my twin they should have most all of these similar interests. These questions really show what I like and integral parts of me and what I value. ✏️)*

<hr style="color: #DD3403;">

# Part 2: The Similarity Function

Now that you have chosen the **questions** you'd like to use to find your DISCOVERY twin, it's time to create a **function** to calculate a similarity value for each person, or each row in `df_hello`.

In lecture, remember learning that a Python **function** takes an input in and returns a **value**. For this specific function, we will take a **row** and calculate a **similarity score** for that row.

But first, **print your list of questions** using the Python cell below for reference:

### Puzzle 2.1: The Similarity Function
We have provided the basic outline for the `similarity` function below. Follow the **commented details** and **complete the function** using all of your questions below:

In [14]:
# we define the function "similarity" to take in a row, stored as the variable "row"

def similarity(row):
    # This function will keep track of a `score` of how similar two rows are with each other.
    # The score begins at zero and we will add to it every time a row matches.
    score = 0 

    # This if statement below checks to see if the `row` shares your choice between Pancakes and Waffles
    # - If the values are the same (==), we add one to the score.
    if row['Android or iOS?'] == my_row['Android or iOS?']: score = score + 1
    if row['Canes or Chick-fil-a?'] == my_row['Canes or Chick-fil-a?']: score = score + 1
    if row['Marvel or DC?'] == my_row['Marvel or DC?']: score = score + 1
    if row['UIUC Football or Basketball Team?'] == my_row['UIUC Football or Basketball Team?']: score = score + 1
    if row['In-State or Out-of-State/International'] == my_row['In-State or Out-of-State/International']: score = score + 1
    if row['Introvert or Extrovert?'] == my_row['Introvert or Extrovert?']: score = score + 1
    if row['Morning Person or Night Person?'] == my_row['Morning Person or Night Person?']: score = score + 1
    if row['Do you like partying at the University of Urbana-Champaign or do you want to go back home?'] == my_row['Do you like partying at the University of Urbana-Champaign or do you want to go back home?']: score = score + 1
    if row['Rap Music or Pop Music?'] == my_row['Rap Music or Pop Music?']: score = score + 1
    if row['Starbucks or Dunkin?'] == my_row['Starbucks or Dunkin?']: score = score + 1
    if row['Spring or Fall?'] == my_row['Spring or Fall?']: score = score + 1
    if row['TV Shows or Movies?'] == my_row['TV Shows or Movies?']: score = score + 1

    # Use this syntax for each question (column) that you have chosen in your list.
    # IMPORTANT: Make sure to keep your lines all at this same level of tabbing out from the start.    
    # Delete or the replace the "Pancakes or Waffles" question, unless you've decided that is one of your evaluating questions. 

    # Finally, return the `score`.
    # ...this line should always be the LAST line.
    return score

### 🔬 Test Case Checkpoint 🔬

In [15]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(similarity(my_row) == len(questions) or similarity(my_row) == my_row[questions].count()), "Ensure you've added all the questions from your list and only your list correctly to your similarity function."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 2.2: Applying the Similarity Function
Now that you've properly defined the similarity function, we need to **apply** it to the rest of the Hello Dataset DataFrame so we can produce a **similarity score** for each person who took the survey!

Luckily, `pandas` provides a useful function for our use case: `df.apply()`.

Quick lesson: `df.apply(some_function)` applies a pre-defined function named `some_function` along an axis of your DataFrame. Because we want to run our `similarity` function on all **rows**, when you call `df.apply()`, ensure you use **`axis=1`**. This would look something like `df.apply(some_function, axis=1)`.

Using the above explanation as reference, **add a column** named `Similarity_Score` to your original DataFrame that contains the value after running your similarity function you defined earlier across all rows:

In [16]:
df_hello['Similarity_Score'] = df_hello.apply(similarity,axis=1)

### 🔬 Test Case Checkpoint 🔬

In [17]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert('Similarity_Score' in df_hello), "Make sure you've named and added the Similarity_Score column to your original df_hello"
assert(df_hello['Similarity_Score'].dtype == 'int64'), "Double check that you've called the similarity function correctly in your df.apply()"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: Finding your Twin(s)

Your last column in `df_hello` is now a metric (a measurement) of how similar you are to everyone else in the class, based on the questions you've chose!

Finally, we can do analytical Data Science! 

### Puzzle 3.1: Your Twin in 107

Using the Python cell below, find the **person most similar to you in the entire course** using the `Similarity_Score`.

Note that **you should be most similar to yourself** (you will always match yourself)! This is not the right answer. Find the person most similar to you *aside from yourself*.

In [27]:
df_twin = df_hello.nlargest(2,"Similarity_Score")
df_twin

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
119  Sabarish  AYD                    2.0                    5.0   
265  Priyansh  AYF                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
119                                          2-Jan                 7UP   
265                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
119             iOS            Apple          Pork   On/Before June 30  ...   
265             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite mobile app? What is your favorite subject?  \
119                         Instagram                           Math   
265                          iMessage                           Math   

    Who is your favorite singer? What is your favorite season?  \
119                   Bruno Mars                          Fall   
265                   Steve Lacy                          Fall   

    What is your Zodiac sign? What is your favorite video game?  \
119                     Virgo                             Forza   
265                 Aquarius                  Call of Duty: BO3   

    What is your favorite non-video game? What is your favorite food?  \
119                              Monopoly               Chicken Wings   
265                              Football                        Thai   

    What's your favorite movie? Similarity_Score  
119         Wolf of Wall Street               12  
265               Interstellar                12  

[2 rows x 121 columns]

### Analysis: True Twin?

**Q: If you picked twelve or more other questions to evaluate similarity, do you think you'd find a different twin in the course overall? Explain why or why not in at least 3 complete sentences.** 

*(✏️ I think that if I picked twelve or more other questions I would have had a different twin. The reason is because we have different answers for most all of the other questions. I still think this is the most accurate because the questions I picked were the most important to me and because we share the most important values, I think we are really similar. I know this kid and he is my friend which is really interesting and funny to me. ✏️)*

### Puzzle 3.2: Your Lab Twin

Now, chances are your twin isn't in your lab section. Because we conveniently have our row, `Similarity_Score` applied to the **entire** set of responses to the survey, we can create smaller dataframes to answer more questions. 

Using the Python cell below, find the person most similar to you in **your lab section**. Creating a new DataFrame might help. 


In [29]:
df_lab = df_hello[df_hello.Lab == "AYD"]
df_labtwin = df_lab.nlargest(2,"Similarity_Score")
df_labtwin

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
119  Sabarish  AYD                    2.0                    5.0   
32     Beomsu  AYD                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
119                                          2-Jan                 7UP   
32                                           2-Jan         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
119             iOS            Apple          Pork   On/Before June 30  ...   
32              iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite mobile app? What is your favorite subject?  \
119                         Instagram                           Math   
32                          Instagram              Computer Science    

    Who is your favorite singer? What is your favorite season?  \
119                   Bruno Mars                          Fall   
32                   Central cee                       Winter    

    What is your Zodiac sign? What is your favorite video game?  \
119                     Virgo                             Forza   
32                     Sheep                          Valorant    

    What is your favorite non-video game? What is your favorite food?  \
119                              Monopoly               Chicken Wings   
32                                  poker                       Pizza   

    What's your favorite movie? Similarity_Score  
119         Wolf of Wall Street               12  
32               Fast & Furious               10  

[2 rows x 121 columns]

#### **Once you have that person, find them!** If they're not here, look for the next similar person.
Help them finish their lab and let them find their DISCOVERY twin!

Note that there may be ties in Similarity_Score due to the usage of only parts of the dataset.

### Puzzle 3.3: The Course Staff
A lot of wonderful course staff work together to bring you all the content for Data Science DISCOVERY. 

The course staff of 107 consists of the **Professors, TAs, and CAs**. They've all completed the Hello Survey too! 

Their response under the `name` column is standardized and follows these formats:
- 'Prof. Name'
- 'TA Name' 
- 'CA Name'

But how can we create a new DataFrame with just the course staff? Python saves the day again.

#### **The `contains` Function**
In Python, the `contains` function can be used to check if a specific **substring** is present in a **string** 

This will come in handy for us, because we can see if 'CA ', 'TA ', or 'Prof. ' is in the **names** of any rows in our DataFrame

Using the `contains` function, create a **new DataFrame** with just the course staff: `df_staff`. 

Part of the statement has been completed for you to show you how to use `contains`. 

*Note that we have to cast our `Name` column of `df_hello` to a **string** by using `.str`

For more information, see https://discovery.cs.illinois.edu/guides/DataFrame-Row-Selection/dataframe-string-contains/

In [31]:
df_staff = df_hello[(df_hello['Name'].str.contains('CA ')) + df_hello['Name'].str.contains('TA ') + df_hello['Name'].str.contains('Prof. ')] 
df_staff
# IMPORTANT: Make sure you include a whitespace in 'CA ' and 'TA ' so that people with those letters in their name aren't included. Whitespace counts. 

Name          Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0         CA Ana          AL1                    2.0                    5.0   
1         CA Ram          AYR                    2.0                    5.0   
2         CA Xin          AYO                    1.0                    5.0   
46    Prof. Wade          AL1                    1.0                    5.0   
48     CA Yehjee          AYK                    1.0                    3.0   
51        CA Mia          AYL                    1.0                    3.0   
56    CA Kaitlyn  AYA and AYC                    2.0                    3.0   
57     CA Eliana     AYM, AYS                    2.0                    5.0   
69       CA Nick          AYB                    1.0                    3.0   
83   Prof. Karle          AL1                    1.0                    3.0   
87   CA Jessica           AYS                    1.0                    5.0   
95     TA Trisha     AYI, AYM                    1.0                    3.0   
98        CA Jai     AYA, AYQ                    1.0                    3.0   
106        CA A🐝      AYE,AYG                    2.0                    3.0   
118    TA Hanjia     AYC, AYE                    1.0                    5.0   
126    TA Tianyi      AYJ AYT                    1.0                    3.0   
145      CA Evan     AYF, AYH                    2.0                    3.0   
151    TA Daniel     AYB, AYD                    1.0                    5.0   
227    CA Shriya     AYP, AYR                    2.0                    3.0   
263     CA Talia          AYH                    2.0                    5.0   
270    CA Alyssa     AYE, AYG                    1.0                    5.0   
356     CA Emily          AYB                    1.0                    3.0   
372      CA Jake          AYC                    1.0                    5.0   
425      TA Urvi     AYH, AYO                    1.0                    3.0   
472      CA Linh     AYD, AYF                    1.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
46                                           2-Jan                 NaN   
48                                             0.5                 7UP   
51                                           2-Jan                 7UP   
56                                             0.5                 7UP   
57                                           2-Jan         Sierra Mist   
69                                             0.5         Sierra Mist   
83                                             0.5                 7UP   
87                                           2-Jan                 7UP   
95                                           2-Jan                 7UP   
98                                           2-Jan         Sierra Mist   
106                                          2-Jan                 7UP   
118                                          2-Jan                 7UP   
126                                          2-Jan                 7UP   
145                                            0.5         Sierra Mist   
151                                          2-Jan         Sierra Mist   
227                                          2-Jan                 7UP   
263                                          2-Jan         Sierra Mist   
270                                          2-Jan         Sierra Mist   
356                                          2-Jan                 7UP   
372                                            0.5         Sierra Mist   
425                                          2-Jan                 7UP   
472                                          2-Jan                 7UP   

    Android or iOS? Apple or Ora

### 🔬 Test Case Checkpoint 🔬

In [32]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df_staff) == 25), "Check you've added all the course staff correctly."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4: Course Staff Twin!

Now that we have the course staff responses in `df_staff`, find your course staff twin (the staff member who has the **highest Similarity_Score** to you)

In [33]:
df_stafftwin = df_staff.nlargest(1,"Similarity_Score")
df_stafftwin

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0  CA Ana  AL1                    2.0                    5.0   

  What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                          2-Jan         Sierra Mist   

  Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0             iOS            Apple          Pork   On/Before June 30  ...   

  What is your favorite mobile app? What is your favorite subject?  \
0                          Snapchat                           Math   

  Who is your favorite singer? What is your favorite season?  \
0                 Taylor Swift                          Fall   

  What is your Zodiac sign? What is your favorite video game?  \
0                    Piceis                         Minecraft   

  What is your favorite non-video game? What is your favorite food?  \
0                    Chutes and Ladders                      Yogurt   

  What's your favorite movie? Similarity_Score  
0                           ?                9  

[1 rows x 121 columns]

### Analysis: Final Reflection

**Q: Do you think our method of finding similar people is effective? Why or why not? Write at least 2 complete sentences.**

*(✏️ I don't think this is the most effective method to find similar people. I think that the questions need to be more specific and relate more to personality and deeper preferences to truly find the best soulmate or similar person.  ✏️)*

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)